### ISSUE TITLE
subseq => core dumped

### ISSUE THREAD

**Issue by ndaniel** [lh3/seqtk#46]
It looks like when there is not enough free memory available for "seqtk subseq", then a crash happens with this error message: "Segmentation fault (core dumped)".

For example running this:

seqtk subseq reads-of-size-35-GB.fq list-of-size-20-GB.txt > output.fq

on a computer with 16GB RAM will produce  this error messge "Segmentation fault (core dumped)".

The expected behavior would be to end graciously with a message like "not enough memory" and so on!

**Comment by tseemann** [comment]
The `stk_reg_read()` function which allocates the hash for the sequence IDs uses the `malloc` and `realloc()` system calls but does not check to see if they failed.

Because you loaded a file bigger than your RAM size, it failed.  The simplest solution is to break your `list.txt` file into smaller pieces and do them one at a time. You can use the Unix `split` command to do that.

**Comment by ndaniel** [comment]
Unfortunately it is not possible to break the 'list.txt' becasue SEQTK is run as part of a automatic pipeline (that is https://github.com/ndaniel/fusioncatcher ) on a server which runs simultaneously several tasks/programs. Therefore the free memory available on the server is changing all the time and it makes it impossible to predict it beforehand if SEQTK has enough free memory or not.

The right way to fix this bug is to check the return code of malloc() and realloc() in SeqTK.

**Comment by tseemann** [comment]
@ndaniel Sure - an error code would be good practice - but fusioncatcher will still fail?

**Comment by ndaniel** [comment]
> @ndaniel Sure - an error code would be good practice - but fusioncatcher will still fail?

Usually SeqTK will become a zombie process and it will refuse in many cases to die for hours (and even days) afterwards when it was not able to allocate memory successfully. Therefore the pipeline which is using it will also hang in a zombie mode for hours and days because of it.

>Another option is to use grep -f -F list.txt -A 3 file.fastq

Also I have had already a replacement in place when SeqTK when it fails, that is https://github.com/ndaniel/fusioncatcher/blob/master/bin/extract_short_reads.py , but this is slower then SeqTK. Therefore there is not need for workarounds.

**Comment by tseemann** [comment]
@ndaniel yes it would be good if it failed correctly so you could fallback onto your python script. but given you know how much RAM you have, can you just fallback when you know the size of the `list.txt` file is too big?

**Comment by ndaniel** [comment]
> but given you know how much RAM you have, can you just fallback when you know the size of the list.txt file is too big?

No. It is impossible to predict how much available free RAM at a given time in the future on a server which runs/starts/stops simultaneously several tasks/programs.

> also, i don't think making the read IDs shorter will help, because the code uses a fixed length hashing string to store the IDs.

Yes, it does help according to my tests.

**Comment by shenwei356** [comment]
>seqtk subseq reads-of-size-35-GB.fq list-of-size-20-GB.txt > output.fq

Since the whole IDs list needs to be stored in RAM, a memory efficient data structure like ***BloomFilter*** could be used for checking the existence of a ID. And *this probabilistic data may has false positives but does not has false negatives*, so we can use the *subtract* (IDs not in sequence file) to ensure the correctness.

This needs more steps:

1. compute the substract
2. construct the bloom filter with substract
3. extract sequences no in the substract

**Comment by ndaniel** [comment]
@shenwei356 

For now I am using this [extract_short_reads.py](https://github.com/ndaniel/fusioncatcher/blob/master/bin/extract_short_reads.py) for cases when the reads ids do not fit in the memory and it works well BUT it is slower than SeqTK.
